ceci/stage.py

```diff
@@ -811,69 +906,86 @@ def read_config(self, args):
+        Parameters
+        ----------
+        global_config : dict
+            Global parameters not to write
+        reduce_config : bool
+            If true, reduce the configuration by parsing out the inputs, outputs and global params
 
-        return my_config
+        Returns
+        -------
+        out_dict : dict
+            The configuration
+        """
+        out_dict = {}
+        if reduce_config:
+            ignore_keys = self.input_tags() + self.output_tags() + ['config']
+        else:
+            ignore_keys = []
+        for key, val in self.config.items():
+            if reduce_config:
+                if key in global_config:
+                    if global_config[key] == val:
+                        continue
+                if key in ignore_keys:
+                    continue
+            out_dict[key] = cast_to_streamable(val)
+        return out_dict
+
+
+    def find_inputs(self, pipeline_files):
+        """Find and retrun all the inputs associated to this stage in the FileManager
+
+        These are returned as a dictionary of tag : path pairs
+        """
+        ret_dict = {}
+        for tag, _ in self.inputs:
+            aliased_tag = self.get_aliased_tag(tag)
+            ret_dict[aliased_tag] = pipeline_files[aliased_tag]
+        return ret_dict
+
+    def find_outputs(self, outdir):
+        """Find and retrun all the outputs associated to this stage
 
-    def iterate_fits(self, tag, hdunum, cols, chunk_rows, parallel=True):
+        These are returned as a dictionary of tag : path pairs
+        """
+        ret_dict = {}
+        for tag, ftype in self.outputs:
+            aliased_tag = self.get_aliased_tag(tag)
+            ret_dict[aliased_tag] = f"{outdir}/{ftype.make_name(aliased_tag)}"
+        return ret_dict
+
+    def print_io(self, stream=sys.stdout):
+        """Print out the tags, paths and types for all the inputs and outputs of this stage"""
+        stream.write("Inputs--------\n")
+        for tag, ftype in self.inputs:
+            aliased_tag = self.get_aliased_tag(tag)
+            stream.write(f"{tag:20} : {aliased_tag:20} :{str(ftype):20} : {self._inputs[tag]}\n")
+        stream.write("Outputs--------\n")
+        for tag, ftype in self.outputs:
+            aliased_tag = self.get_aliased_tag(tag)
+            stream.write(f"{tag:20} : {aliased_tag:20} :{str(ftype):20} : {self._outputs[aliased_tag]}\n")
+
+    def should_skip(self, run_config):
+        """Return true if we should skip a stage b/c it's outputs already exist and we are in resume mode"""
+        outputs = self.find_outputs(run_config["output_dir"]).values()
+        already_run_stage = all(os.path.exists(output) for output in outputs)
+        return already_run_stage and run_config["resume"]
+
+    def already_finished(self):
+        """Print a warning that a stage is being skipped"""
+        print(f"Skipping stage {self.instance_name} because its outputs exist already")
```

Comment:
I'm not sure when this should be printed - right now it prints when you create the pipeline.

Reply:
Hmm. good question.  It is getting printed in `initialize` (was `load_configs`) which is getting called `create` which is called from `read`, but not in the c'tor of Pipeline.   So it gets called when you read an existing pipeline, or if you happened to `create` one from a dictionary of configuration parameters, but will not be called in you make one explicitly.  

I think that this is probably ok.   If you are reading a pipeline where stuff has already been created, might make sense to let people know ASAP.